Cleanup patch for TLuaInterpreter

smurfix opened the pull request:
Remove two anti-patterns from TLuaInterpreter.cpp:

* De-indent the "else" branch in this code::
```
    if(error) {
        ...
        lua_error(L);
        return 1;
    } else {
        CODE
    }
```
=>
```
    if(error) {
        ...
        lua_error(L);
        return 1;
    }
    CODE
```
* Also, move variable declarations down to keep them with the code
  initializing them, where that makes sense::
```
    int foo;
    if(error) {
        ...
    } else {
        foo = lua_tointeger(L, 1)
    }
```
=>
```
    if(error) {
        ...
    }
    int foo = lua_tointeger(L, 1)
```
* Do not create std::string variables only to convert them to a QString later.

The first+second change has been requested by @SlySven in pr#4003.

smurfix committed:
Cleanup

Remove two anti-patterns from TLuaInterpreter.cpp:

* De-indent the "else" branch in this code::

    if(error) {
	    ...
		lua_error(L);
	    return 1;
    } else {
	    CODE
    }

=>

    if(error) {
	    ...
		lua_error(L);
	    return 1;
    }
	CODE

* Also, move variable declarations down to keep them with the code
  initializing them, where that makes sense::

    int foo;
    if(error) {
	    ...
    } else {
	    foo = lua_tointeger(L, 1)
    }

=>

    if(error) {
	    ...
    }
	int foo = lua_tointeger(L, 1)

* Do not create std::string variables only to convert them to a QString later.

The first+second change has been requested by @SlySven in pr#4003.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -1962,9 +1918,8 @@ int TLuaInterpreter::getLineCount(lua_State* L)
-        lua_pushnumber(L, mudlet::self()->getLastLineNumber(&host, _window) + 1);
+        QString window = lua_tostring(L, 1);
+        lua_pushnumber(L, mudlet::self()->getLastLineNumber(&host, window) + 1);
         return 1;
     } else {
         int lineNumber = host.mpConsole->getLineCount();
```
Here's another one.
Also the final `return 0;` seems obsolete

smurfix replied:
I agree about the final "return 0" there; will update the patch.

However I intentionally didn't touch any cases where the "return" does not precede a call to `lua_error`. The reason is that an error call is not symmetric, but some (most?) of the other tests are.

However, there *are* many `return`s which are preceded by "push NIL and an error message" which should be treated like an error; I'll do that also.

Edru2 replied:
> However, there are many returns which are preceded by "push NIL and an error message" which should be treated like an error; I'll do that also.

I think this is intentional so the script in lua doesn't break. (I could be wrong)

vadi2 replied:
Correct, API design is:

* wrong arguments: error
* correct arguments, but a problem executing it: return nil+msg, do not error

This is standard Lua design

Kebap replied:
Not sure how this error handling influences the if-return-else antipattern.
For me, the else seems superfluous still.

vadi2 replied:
I agree, the else is superfluous.

smurfix replied:
Well, `if(foo) return; else bar()` is always equivalent to `if(foo) return; bar()`. That's not the question.

My point is, there's a conceptual difference between returning an error and returning different values depending on the function's arguments. In the error cases (both `lua_error` and `nil-plus-errstr`), the `bar()` should not be indented. However, when we return different values both branches should be recognizeable as equivalent (thus you can swap the branches or add another clause without affecting the indentation), so in that case we should keep the explicit `else`.

SlySven replied:
The `return 0;` is unnecessary - once the `else-after-return` is cleaned up it will be even more clear that it is unreachable code...

On the other hand there is a glaring omission of code (and documentation in the wiki) to handle the case where the `window` argument is supplied but does not exist. Our current paradigm is for the `(int) mudlet::getLastLineNumber(Host*, QString&)` to be revised to return a `std::pair<bool, QString>` where, if there is an error in the deeper code the first argument is a `false` and *in that case* the second argument is a non-translatable string (so will be generated from a `QStringLiteral` which contains a message that may include the `window` value generated in the deeper code) which is then pushed onto the Lua stack preceded with a `nil`. As the `window` name need not contain a Latin1 string we have to use the `QString::toUtf8()` method on the string before we can squirt it back to the Lua subsystem.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -2568,42 +2517,37 @@ int TLuaInterpreter::moveCursor(lua_State* L)
     Host& host = getHostFromLua(L);
 
-    if (a1 == "main" || n < 3) {
+    if (windowName.isEmpty() || windowName == "main") {
```
Here you drop the `n < 3` clause, but I think it is alike to `windowName.isEmpty()` - maybe this even improves things

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -2614,42 +2558,37 @@ int TLuaInterpreter::setConsoleBufferSize(lua_State* L)
     Host& host = getHostFromLua(L);
 
-    if (a1 == "main" || n < 3) {
+    if (windowName.isEmpty() || windowName == "main" || n < 3) {
```
Here you keep the `n < 3` clause. Was this by accident or am I missing some other difference?

smurfix replied:
No, I missed that one.
Anyway it seems that I also need to go through with another fine-toothed comb and fix all the `empty name == missing name == name='main'` cases so it's remotely consistent. Will be in the next update.

smurfix committed:
superflous "return"

smurfix committed:
Also treat pushnil+pushstring+return as error

i.e. de-indent the "else" part.

I intentionally kept structures like

    if (bad1) {
		// push error one and return
	} else if (bad2) {
		// push error two and return
	}

the way they are, even though the "else" is superfluous, because this
keeps the error handling block visually coherent; otherwise it'd become
a line longer.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -2742,61 +2677,54 @@ int TLuaInterpreter::disableCommandLine(lua_State* L)
     Host& host = getHostFromLua(L);
     if (n == 1) {
-        host.mpConsole->replace(QString(a1.c_str()));
+        host.mpConsole->replace(text);
```
Whoops, shouldn't this be `name` instead..!?

Your named variables are a bit confusing, because argument #1 can be optional.
If there is only one argument, then it will be text, otherwise the window name.
Maybe use a name like `name_or_text` for the first argument instead?

smurfix replied:
Thans. Will fix.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -2742,61 +2677,54 @@ int TLuaInterpreter::disableCommandLine(lua_State* L)
     Host& host = getHostFromLua(L);
 
-    if (name.empty()) {
+    if (name.isEmpty() || name == "main") {
```
Fair addition, please update documentation!

smurfix replied:
In fact I'm going through this and fix the test everywhere. There's a fair amount of "main" tests both in Lua and in TConsole.cpp that are duplicate and/or inconsistent.

smurfix replied:
As to the Wiki docs, I'll update it when this gets merged.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -3129,9 +3043,8 @@ int TLuaInterpreter::loadMap(lua_State* L)
-        } else {
-            location = QString::fromUtf8(lua_tostring(L, 1));
         }
+        location = QString::fromUtf8(lua_tostring(L, 1));
```
```suggestion
        QString location = QString::fromUtf8(lua_tostring(L, 1));
```
Then you can again delete the `QString location;` line above as well.

Kebap replied:
Nevermind, things aren't that easy here.

smurfix committed:
Fixed a variable naming error

smurfix committed:
TLuaInterpreter: Collect "main" window name tests

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -92,6 +92,15 @@ int speechState = QTextToSpeech::Ready;
 
+static bool isMain(const QString& name)
+{
+    if (name.isEmpty())
```
Please include the `{`...`}` even for single lines of code.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -92,6 +92,15 @@ int speechState = QTextToSpeech::Ready;
+{
+    if (name.isEmpty())
+        return true;
+    if (name.compare(QStringLiteral("main"), Qt::CaseSensitive) == 0)
```
I've mentioned this in another PR recently but I'm not sure if they are both by the same person so I will repeat myself, in my *humble* opinion (unless @vadi2 thinks otherwise) these lines should be changed to:
```cpp
     if (!name.compare(QStringLiteral("main"))) {
```
The logic of comparing a function result as equal or not equal to `0` in an `if` is confusing as `0` is the same as `false` in C/C++ - so this test sort of simplifies to `if ( booleanReturningFunction() == false )` - and `if (!booleanReturningFunction())` is more concise...

... also for `(int) QString::compare(...)`: `Qt::CaseSensitive` is the default anyway - you would only need to put a second argument in the opposite {`Qt::CaseInsensitive`} was needed!

vadi2 replied:
Looks good to me too. We just used the other pattern but if this works, it's much cleaner!

smurfix replied:
Well, if we want "clean" then surely `name == "main"` is the cleanest of them all. *g* but I'll change it to this one.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -362,14 +371,13 @@ int TLuaInterpreter::Wait(lua_State* L)
-    } else {
-        luaSleepMsec = lua_tointeger(L, 1);
     }
+    int luaSleepMsec = lua_tointeger(L, 1);
```
👍

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -653,45 +661,35 @@ int TLuaInterpreter::selectString(lua_State* L)
     if (lua_gettop(L) > 2) {
         if (!lua_isstring(L, s)) {
             lua_pushfstring(L, R"(selectString: bad argument #%d type (window name as string, is optional {defaults to "main" if omitted}, got %s!))", s, luaL_typename(L, s));
             lua_error(L);
```
BTW, a further clean up we want to back-port into some of the older functions is to change:
```cpp
    lua_error(L);
    return 1;
```
to
```cpp
    return lua_error(L);
```

`lua_error(L)` calls **[`longjmp()`](https://en.wikipedia.org/wiki/Longjmp)** and that ***never returns*** so that the following `return 1;` is never reached - but putting it in helps with some code analysis programs (and helps to remind the developer that it does represent an exit point {albeit non-standard} from **this** function).

Kebap replied:
Out of scope of this PR here... Maybe file a new issue?

smurfix replied:
I'll keep that in the back of my mind along with the other points raised during this discussion. There definitely will be a follow-up PR or two.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -702,35 +700,34 @@ int TLuaInterpreter::selectString(lua_State* L)
-    if (lua_gettop(L) == 0) {
-        luaSendText = "main";
-    } else {
+    std::string windowName;
```
💡 Eventually we want to standardise things to make the `TConsole` methods all take a `QString&` as the name rather than a `std::string`.

smurfix replied:
That's another fun design issue. Right now the calls to find the correct subconsole are nicely spread between mudlet.cpp, TConsole.cpp, and one call in TLuaInterpreter. That's a lot of code, all of which does the exact same thing.

Why do we need that?

I'd much rather have a single `getNamedConsole(windowName)` method in TConsole that returns `this` if the name indicates the main console, `null` if it can't be found, and the named sub-`TConsole` instance otherwise. That'd allow us to remove all `windowName` parameters, as well as all sub-console-dispatch functions from mudlet.cpp.

SlySven replied:
Actually the main `TConsole` is actually the entry-point into access into the profile's other `TConsole` or at least the ones that the user creates and interacts with (Oh! ***And*** the `TLabel`s). The proposal to have a `getNamedConsole(...)` is sort of contrary to the C++ idea of encapsulation - the Lua system should not be poking around in the depths of the `TConsole` class, instead it should pass a message about what it wants to a single instance and that should handle the details...

At the moment, though, there are an unhelpfully large number of functions (probably dating back to a time when Mudlet could only open one profile at a time ???) that instead go via the `mudlet` singleton class (specifying a `Host` AND a console name) that end up being routed to the Main `TConsole` anyhow. {The reason being that that console marshals the others because it holds the list of pointers to the other ones - and because some of the other - the sub-consoles - are contained inside it and when it is destroyed they must be as well.}

The bottom line is that we should be aiming to go to the main `TConsole` (which *should* always be present) for all console related functions and it should either do what is asked for, either to itself, or to one of the other class instances that it controls, and return either the wanted data - if appropriate or a success or failure indication, and if there is more than one failure cause it should indicate which.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -702,35 +700,34 @@ int TLuaInterpreter::selectString(lua_State* L)
     std::string windowName = "main";
 
     if (!lua_isnumber(L, 1)) {
-        lua_pushfstring(L, "isAnsiFgColor: bad argument #1 type (window name as string expected, got %s!)", luaL_typename(L, 1));
```
🔎 Well spotted! 😀

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -967,7 +959,7 @@ int TLuaInterpreter::getBgColor(lua_State* L)
 int TLuaInterpreter::getTextFormat(lua_State* L)
 {
-    QString windowName = QStringLiteral("main");
+    QString windowName;
```
Not sure about this change - if NO arguments are supplied then `windowName` will be left as a `QString()` i.e. a null one.

smurfix replied:
Doesn't change things because TConsole::getTextAttributes also checks for both empty and "main".

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -1962,16 +1924,14 @@ int TLuaInterpreter::getLineCount(lua_State* L)
-        std::string window = lua_tostring(L, 1);
-        QString _window = window.c_str();
-        lua_pushnumber(L, mudlet::self()->getLastLineNumber(&host, _window) + 1);
+        QString window = lua_tostring(L, 1);
```
Need to convert it from UTF-8 - should be:
```cpp
QString window = QString::fromUtf8(lua_tostring(L, 1));
```
or even more concisely (and making it `const` to help the compiler in opimisations):
```cpp
const QString window{QString::fromUtf8(lua_tostring(L, 1))};
```

smurfix replied:
This may be a stupid question, but why would we need QString::fromUtf8 here? Or in fact anywhere?

The QString documentation clearly states that QString::fromUtf8 is implicit in constructors from `const char*` (which includes assignment AFAICT).

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -2742,61 +2682,55 @@ int TLuaInterpreter::disableCommandLine(lua_State* L)
-        a1 = lua_tostring(L, s);
-        s++;
     }
+    QString text = lua_tostring(L, s);
```
Need `QString::fromUtf8(...)` here...

SlySven replied:
... actually you will need this in *quite a few* places!

vadi2 replied:
Actually, no:

![image](https://user-images.githubusercontent.com/110988/95015335-914dac80-064c-11eb-88e6-2c5328334dac.png)

As @smurfix pointed out, we are needlessly using it everywhere, because Qt already uses internally: https://doc.qt.io/qt-5/qstring.html#QString-7

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -2742,61 +2682,55 @@ int TLuaInterpreter::disableCommandLine(lua_State* L)
-            a2 = lua_tostring(L, s);
         }
+        windowName = text;
+        text = lua_tostring(L, s);
```
And here.

SlySven reviewed (commented):
Only got as far as getExitStubs()

It might be that this PR is too large - yeah, I know, coming from ME that is a bit rich! 🙈

smurfix commented:
@SlySven I don't think the PR is too large, but by now it's in danger trying to fix too much at once. You added comments that opened two other cans of worms.

While I mostly agree with all of them, I would prefer to fix the "obvious" errors in this PR, get it merged, and then tackle the other problems one-by-one. Otherwise we'll get lost in a maze of changes which fix three things at once, which invariably means that some of then will only fix two, which will bite somebody later.

vadi2 commented:
I agree, let's scope the list of issues to what we have here now, other improvements can be done as a follow up.

smurfix committed:
Prettify isMain

* doc comment that it's undocumented ;-)
* add braces
* use !QString::compare instead of ==0
* drop the CaseSensitive, it's the default

smurfix committed:
be consistent

at least with the line before this

smurfix committed:
Replace all other instances of `== "main"`

with calls to `isMain()`

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -92,12 +92,15 @@ int speechState = QTextToSpeech::Ready;
-    if (name.compare(QStringLiteral("main"), Qt::CaseSensitive) == 0)
+    }
+    if (!name.compare(QStringLiteral("main"))) {
         return true;
```
This is a new antipattern, if something return true else return false.
Why not just return something?

smurfix replied:
Mainly because QString::compare is zero when it matches, which (if you're testing for equality) is also an antipattern, burned into the brains of every C programmer for the last half century. That doesn't make it any more correct, thus I really try to avoid directly returning that.

Translated into plain English, "if it's empty, return True. If it's `main`, return True. Otherwise return False" is immediately understandable. "if it's empty, return True, otherwise return whether it … umm … equals `main`" somewhat less so.

Alternately of course we could rewrite that test as `name == "main"` …

smurfix commented:
IMHO this is ready for merging now. Let's create issues for the other unresolved problems this patch doesn't address.

Kebap commented:
I am still reviewing this, currently at line 6000 / 18000, but it's taking some time. 

edit: 12500
edit: done

smurfix commented:
OK, no hurry, feel free to push fixes on top if you find anything I overlooked.

vadi2 commented:
I need to sit down to review as well, will do soon.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -10244,72 +9819,70 @@ int TLuaInterpreter::setDoor(lua_State* L)
-            }
-            // else IS a valid stub or real normal exit -fall through to continue
+    }
+    exitCmd = QString::fromUtf8(lua_tostring(L, 2));
```
```suggestion
    QString exitCmd = QString::fromUtf8(lua_tostring(L, 2));
```
and drop the line above

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -10320,16 +9893,15 @@ int TLuaInterpreter::setDoor(lua_State* L)
-            return 2;
-        }
+    }
+    doorStatus = lua_tointeger(L, 3);
```
drop the declaration above and
```suggestion
    int doorStatus = lua_tointeger(L, 3);
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -11239,35 +10775,32 @@ int TLuaInterpreter::clearMapUserDataItem(lua_State* L)
-            return 2;
-        } else {
-            lua_pushboolean(L, (host.mpMap->mUserData.remove(key) > 0));
-            return 1;
-        }
+        lua_pushboolean(L, (host.mpMap->mUserData.remove(key) > 0));
```
Another superfluous else?

Kebap reviewed (approved):
For me, this seems good then. Pending Vadi/SlySven's approval.

smurfix commented:
OK. I'll resolve your remaining comments.

Edit: done

smurfix committed:
Missed a couple of instances WRT previous commits

vadi2 committed:
Merge remote-tracking branch 'origin/development' into cleanup

vadi2 reviewed (approved):
This was a lot of work! Thanks, @smurfix
